tests/pytests/unit/utils/test_dynamic_dict.py

```diff
@@ -0,0 +1,193 @@
+        del copied_ddict[key]
+    assert (
+        not copied_ddict
+    ), "Failed to iterate across all keys in DynamicDict - remaining: {}".format(
+        copied_ddict
+    )
+
+    ddict.add_dyn("foo", dyn_func)
+    copied_ddict = base_dict.copy()
+    copied_ddict["foo"] = dyn_func()
+    for key in ddict.keys():
+        assert key in copied_ddict, f"Found an unexpected key: {key}"
+        del copied_ddict[key]
+    assert (
+        not copied_ddict
+    ), "Failed to iterate across all keys in DynamicDict - remaining: {}".format(
+        copied_ddict
+    )
+
+
+def test_copy(ddict, base_dict, dyn_func):
+    ddict.add_dyn("foo", dyn_func)
+    ddict2 = ddict.copy()
+    copied_dict = base_dict.copy()
+    copied_dict["foo"] = dyn_func()
+    for key in ddict2.keys():
+        assert key in copied_dict, f"Found an unexpected key in the copy: {key}"
+        del copied_dict[key]
+    assert (
+        not copied_dict
+    ), "Failed to iterate across all keys in DynamicDict copy - remaining: {}".format(
+        copied_dict
+    )
+    assert hasattr(
+        ddict2._func_dict.get("foo"), "__call__"
+    ), "Failed to copy dynamic key"
+
+    ddict2 = {"ddict": ddict}
+    copied_dict = {"ddict": base_dict.copy()}
+    copied_dict["ddict"]["foo"] = dyn_func()
+    ddict3 = copy.deepcopy(ddict2)
+    assert ddict3["ddict"].is_dyn_key("foo"), "Dyn key 'foo' is no longer a dynamic key"
+    assert hasattr(
+        ddict3["ddict"]._func_dict.get("foo"), "__call__"
+    ), "Failed to copy dynamic key"
+
+    copied_dict = base_dict.copy()
+    copied_dict["foo"] = dyn_func()
+    for key, val in ddict.items():
+        assert key in copied_dict, f"Found an unexpected key: {key}"
+        del copied_dict[key]
+    assert (
+        not copied_dict
+    ), "Failed to iterate across all keys in DynamicDict - remaining: {}".format(
+        copied_dict
+    )
+
+
+def test_static_dict(ddict, base_dict, dyn_func):
+    ddict.add_dyn("foo", dyn_func)
+    static_dict = ddict.static_dict()
+    assert not isinstance(static_dict, DynamicDict)
+    assert set(static_dict.keys()) == set(ddict.keys())
+    for key, val in static_dict.items():
+        assert val == ddict[key]
+
+
+def test_copy_iterator(base_dict, ddict, dyn_func):
+    ddict.add_dyn("foo", dyn_func)
+    copied_dict = base_dict.copy()
+    copied_dict["foo"] = dyn_func()
+    xvals = [val for val in copied_dict.values()]
+    for val in ddict.values():
+        assert val in xvals, f"Found an unexpected value: {val}"
+        xvals.remove(val)
+    assert (
+        not xvals
+    ), "Failed to iterate across all values in DynamicDict - remaining: {}".format(
+        xvals
+    )
+
+
+def _dyn_func(data=None, dyn_dict=None, key=None):
+    return 7
+
+
+def test_delete_dyn_key(ddict):
+    ddict.add_dyn("foo", _dyn_func)
+    val = ddict.pop("foo")
+    assert val == _dyn_func(), f"Failed pop(): {val} != {dyn_func()}"
```

Comment:
Looks like this is causing the Lint failure:
```
tests/pytests/unit/utils/test_dynamic_dict.py:188: [E0602(undefined-variable), test_delete_dyn_key] Undefined variable 'dyn_func'
```

Reply:
Thanks! Forgot to check back on this and suddenly I can no longer run linting locally with pre-commit, still gotta figure that out... but this should be fixed.